File: FinalProject/Code/Dotnet/BankingSystem/Service/LoanService.cs
using AutoMapper;
using DTO;
using Interfaces;
using Microsoft.EntityFrameworkCore;
using Model;
using Model.DTOs;
using MyDbContext;

namespace Service;

public class LoanService : ILoanService
{
    private readonly MyAppDbContext context;
    private readonly IMapper mapper;
    public LoanService(MyAppDbContext context, IMapper mapper)
    {
        this.context = context;
        this.mapper = mapper;
    }

    public async Task<string> CreateLoanRequestAsync(CreateLoanDTO createLoanDTO)
    {
        try
        {
            var userExists = await context.DbUsers.AnyAsync(u => u.UserId == createLoanDTO.UserId);
            if (!userExists)
                return "User not found.";

            var loanTypeExists = await context.DbLoanType.AnyAsync(l => l.LoanTypeId == createLoanDTO.LoanTypeId);
            if (!loanTypeExists)
                return "Invalid loan type.";

            var loan = new LoanModel
            {
                UserId = createLoanDTO.UserId,
                LoanTypeId = createLoanDTO.LoanTypeId,
                CurrentSalaryInLPA = createLoanDTO.CurrentSalaryInLPA,
                LoanAmount = createLoanDTO.LoanAmount,
                CreatedAt = IndianTime.GetIndianTime(),
                IsApproved = false
            };

            await context.DbLoan.AddAsync(loan);
            await context.SaveChangesAsync();

            return "Loan request created successfully. Awaiting approval.";
        }
        catch (Exception ex)
        {
            Console.WriteLine($"[LoanService] CreateLoanRequest failed: {ex.Message}");
            return "Failed to create loan request.";
        }
    }
    public async Task<string> ApproveLoanAsync(int loanId, int staffId, bool isApproved,string reason)
    {
        using var transaction = await context.Database.BeginTransactionAsync();
        try
        {
            var loan = await context.DbLoan
                .Include(l => l.User)
                .Include(l => l.LoanType)
                .FirstOrDefaultAsync(l => l.Id == loanId);

            if (loan == null)
                return "Loan not found.";

            if (loan.IsApproved)
                return "Loan already processed.";

            loan.IsApproved = isApproved;
            loan.ApprovedBy = staffId;
            loan.RejectionReason = reason;
            loan.ApprovedAt = IndianTime.GetIndianTime();
            loan.isProcessed = true;
            context.DbLoan.Update(loan);
            await context.SaveChangesAsync();
            await transaction.CommitAsync();

            return isApproved ? "Loan approved successfully." : "Loan rejected.";
        }
        catch (Exception ex)
        {
            await transaction.RollbackAsync();
            return $"Approval failed: {ex.Message}";
        }
    }

    public async Task<List<LoanDTO>> GetAllLoansAsync()
    {
        var result = await context.DbLoan
            .Include(l => l.User).ThenInclude(l => l.CustomerType)
            .Include(l => l.LoanType)
            .OrderByDescending(l => l.CreatedAt)
            .ToListAsync();

        return mapper.Map<List<LoanDTO>>(result);

    }
    public async Task<List<LoanDTO>> GetAllLoansByUserAsync(int userId)
    {
        var result = await context.DbLoan
            .Include(l => l.User).ThenInclude(l => l.CustomerType)
            .Include(l => l.LoanType)
            .Where(u=>u.UserId == userId)
            .OrderByDescending(l => l.CreatedAt)
            .ToListAsync();

        return mapper.Map<List<LoanDTO>>(result);

    }

    public async Task<List<LoanDTO>> GetAllLoansToApproveAsync()
    {
        var result = await context.DbLoan
            .Include(l => l.User).ThenInclude(l => l.CustomerType)
            .Include(l => l.LoanType)
            .Where(l => !l.IsApproved  && !l.isProcessed)
            .OrderBy(l => l.CreatedAt)
            .ToListAsync();
        return mapper.Map<List<LoanDTO>>(result);
    }
    
}

File: FinalProject/Code/Angular/BankingSystem/src/app/services/LoanService/loan.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';

export interface CreateLoanDTO {
  userId: number;
  loanTypeId: number;
  loanAmount: number;
  currentSalaryInLPA: number;
}

export interface LoanDTO {
  id: number;
  userId: number;
  userName?: string | null;
  loanTypeId: number;
  loanAmount:number;
  loanTypeName?: string | null;
  createdAt: string;
  isApproved: boolean;
  approvedAt?: string | null;
  approvedBy: number;
  currentSalaryInLPA: number;
  customerType?: string | null;
  isEmployed: boolean;
  rejectionReason:string;
  isProcessed:boolean;
}


@Injectable({ providedIn: 'root' })
export class LoanService {
  private http = inject(HttpClient);
  private base = 'http://localhost:5139/api/v1/Loan';

  create(dto: CreateLoanDTO): Observable<string> {
    return this.http.post(`${this.base}/create`, dto, { responseType: 'text' });
  }
  getMyLoans(userId: number) {
  const params = new HttpParams().set('UserId', String(userId));
  return this.http.get<LoanDTO[]>(`${this.base}/GetMyLoans`, { params });
}
  getPendingApprovals(): Observable<LoanDTO[]> {
    return this.http.get<LoanDTO[]>(`${this.base}/pending-approvals`);
  }
    approveLoan(loanId: number, staffId: number, isApproved: boolean,reason:string): Observable<string> {
    const params = new HttpParams()
      .set('loanId', String(loanId))
      .set('staffId', String(staffId))
      .set('reason', String(reason))
      .set('isApproved', String(isApproved));
    return this.http.post(`${this.base}/approve`, null, { responseType: 'text', params });
  }
    getAllLoans(): Observable<LoanDTO[]> {
    return this.http.get<LoanDTO[]>(`${this.base}/GetAllLoans`);
  }
}
